### DISCUSSION TITLE
Bevy strategy pre-planning for 0.7

### DISCUSSION THREAD

**Discussion by alice-i-cecile** [bevyengine/bevy#2799] [General]
As we wrap up the main work of 0.6 (see #2535 for progress), we're starting to approach the time where we take a look at priorities going forward 🎉

The goal of this discussion is not to *decide* all these things now, but instead to create an agenda and meeting strategy so we can have a proper conversation despite the large number of interested participants :) If there are subtopics here that you would like to be considered that I have not covered, please comment in the appropriate thread and I'll edit it into the parent reply to keep track of topics to discuss. 

There's quite a few things to cover, so I'm going to split this into top-level replies by topic:

- META: Discussion format
- PLANNING: Work to complete before 0.6 launch
- PLANNING: Focus area(s) for 0.7
- PLANNING: Engine development process improvements

**Comment by alice-i-cecile** [comment]
# Discussion format

There are several questions to answer here:

- Scheduled voice chat, Discord chat, or Github Discussions?
- How many meetings?
- If voice or Discord chat, how can we moderate this to reduce chaos?

For those new to the dev-side of the community, decisions are ultimately made by @cart in a fashion that's responsive to community feedback and capacity. If you have comments, concerns or ideas, please chime in!

**Reply by alice-i-cecile** [reply]
I feel pretty strongly that we need 3 separate meetings or discussions to avoid this dragging on forever:
1. 0.6 wrap-up
2. process improvements
3. 0.7 focus areas

I expect that 3 will have the most participation.

**Reply by RustyStriker** [reply]
if you schedule a voice meeting(or any kind of not too long of a meeting) timezones will pose a big problem. So instead of making the "community vote" in the meeting itself, maybe someone can sum it up(or a bunch of people) and post a poll so more people could voice themselves(esp for the 0.7 meeting)

**Reply by afonsolage** [reply]
But discussing it using Discord or GH Discussion will lead to tons of texts, which makes me wonder: The main goal is to discuss some points between the community and some core/main contributors and Cart, or is to generate rich docs about the discussion and let more people interact with it?

If the focus is to have a better and deep understanding between some main people, the meeting is better and we can even record and publish it on YT.

But if the focus is to have a open and broad discussion (like we already have for some topics) so the text channels are better (I think GH discussion should be prioritized).

**Reply by alice-i-cecile** [reply]
> If the focus is to have a better and deep understanding between some main people, the meeting is better and we can even record and publish it on YT.

The challenge with this is that we have core contributors across all time zones. We also have spoken English proficiency / comfort to worry about.

**Reply by Nilirad** [reply]
Unfortunately real time communication via voice chat or Discord threads pose various problems, first of all timezone difference. On the other hand these formats are very useful to exchange information in a very short amount of time. Therefore I think that a hybrid approach may be appropriate.

I see it this way:

1. A preliminary GitHub discussion that poses an issue where the opinion of the community is needed. This discussion will be relevant for at least a couple of days to give as many people as possible time to express their thoughts on the matter.
2. A cart-timezone-centered real time meeting. Here collaborators present to cart a digest of the preliminary discussion. From this, further discussion with present devs is made and a final decision is made. Since some devs cannot attend, they can lay out their submissions in detail in the preliminary discussion and charge who will be present to take their place. 
3. The final decision is then laid out as one or more RFCs (if applicable) and on the #announcements-dev Discord channel and bevy blog.

Note: Discussion material should be archived, linked to or tagged in a way that it is easier to find later.

Note 2: Maybe a short meeting to decide the argument of the preliminary discussion and the subsequent meeting can be done. Or maybe not.

Note 3: For the sake of order, if the meeting is made in the form of voice/video chat, it's better to limit the amount of people who can talk at the same time (e.g. by taking turns or by "raising hand"to comment). For Discord chats, slow mode can be applied, except for people charged to speak.

**Reply by Supermath101** [reply]
I have an idea: What if we can have a "Regional Meetings Moderator/Coordinator" for most major timezones. The idea would be to have multiple meetings, a couple per region based on timezone. Then, record the meetings, and discuss what was talked about on the previous meeting during the next one. Finally, after a few days of cycling around the world having meetings in different timezones, we'll hopefully have reached a global consensus on any major decisions needing to be made.

Note: I already [asked this on Discord](https://discord.com/channels/691052431525675048/745355529777315850/891411523879202846), but I feel it needs to be more formally discussed and recorded. Also, sorry if that sounded a little confusing, I'll be happy to clarify.

**Comment by alice-i-cecile** [comment]
# Work to complete before 0.6 launch

Current things that we know must be completed:

1. #2535
2. #2254
3. Migration guide (https://github.com/bevyengine/bevy-website/pull/185)
4. [Release checklist](https://github.com/bevyengine/bevy/blob/cf221f9659127427c99d621b76c8085c4860e2ef/docs/release_checklist.md).
5. Other items from https://github.com/bevyengine/bevy/milestone/3


The following PRs are things that I think we should consider adding to the milestone:

1. New book (https://github.com/bevyengine/bevy-website/pull/176 and friends)
1. #2765
1. #2736, blocked on #2777
1. #2713
1. #2381
1. #2365

When selecting this list I tried to prioritize a combination of:
- high user impact
- heavily breaking changes
- nearly ready to merge
- not likely to be immediately obsoleted

If there are other open PRs / issues that you think are worth including here, please mention them below. We should also aim to do a quick review (and hopefully merge) of all open `ready-for-final-review` PRs to avoid small fixes going stale.

**Reply by TheRawMeatball** [reply]
Maybe fill out this list using https://github.com/bevyengine/bevy/milestone/3 ? The list is missing a couple important items, namely `#[derive(Component)]`

**Reply by alice-i-cecile** [reply]
Right, good call. Most of the remaining items on the list are small-but-important fixes that I don't think need to be discussed.

**Reply by alice-i-cecile** [reply]
There we go; splitting the "stuff we've already agreed on" from "stuff we need to debate" is a much cleaner organization.

**Reply by hymm** [reply]
should 2777 be on this list too?  I thought it was blocking 2736.

**Reply by alice-i-cecile** [reply]
IMO #2777 is too complex to push for in time, has no direct end-user benefit and is overall too tied up in the broader work being planned #2801. That said, the [judgement call](https://discord.com/channels/691052431525675048/882608549186924554/884873017585455104) from @cart and @Ratysz was that #2777 should be done before #2736, like you said.

So I think that we can rule out #2736 for now, as nice as it would be to have :(

**Reply by TheRawMeatball** [reply]
Eh, I think there's hope for #2777. It's a purely internal change with clear design, and #2736 could be considered it's "direct end-user benefit".

**Comment by alice-i-cecile** [comment]
# Focus area(s) for 0.7

There are several candidates that have been discussed:

- UI
  - needs heavy-handed decision-making to cut through paralysis
  - some serious open design questions, particularly around the data flow model
  - heavy user impact: this is one of the most serious challenges with making Bevy games right now
- System schedule and states (follow #2801 for this planning)
  - move to a stageless design
  - clean up tangle of abstractions
  - simplify States to provide a better user experience
  - fix the "states don't play nice with other run criteria" issues
  - requires a large degree of coordination
  - more system scheduling API improvements: `as_if_before` from #2747, extended variants of system graphs from #2381
  - not very user-visible :(
  - may be able to complete in parallel with another focus area (led by @Ratysz and @TheRawMeatball)
- Scenes and reflection
  - currently limited and underdocumented
  - critical if we want a data-driven UI design
- Assets
  - Distill integration is finally unblocked
  - this area requires lots of expert feedback
  - asset pipelines are important and missing
- Hierarchy clean-up (see #2572)
  - slow, buggy and full of rough edges
  - cannot easily be extended
  - could prioritize incremental improvements before [relations](https://github.com/bevyengine/rfcs/pull/18) prototypes are complete
- Ray-casting, picking and geometric primitives
  - solidly useful, well-scoped
  - critical for potential physics and advanced rendering topics
  - will lean heavily on prototypes and expertise by @aevyrie
  - could include Transform-adjacent topics like #1680, splitting `Transform` or changing the coordinate directions
- Canvas API
  - Very useful for prototyping and making 2D games
  - Solid third-party crate exists: reduces need to prioritize, but also reduces uncertainty
  - Could tie in wire-frame / debug-lines to improve visual prototyping / debugging further
  - Some elements of the UI (e.g. control over screen vs world coordinates) could be tackled incrementally to help this

There are some other areas that I don't think are appropriate candidates at this stage:
- Relations
  - fundamentally a research problem still; dedicated collaboration won't help this at this stage
  -  we need a solid prototype before moving forward
- Audio
  - our audio is badly lacking
  - `bevy_kira_audio` is a solid stop-gap for now; we should watch to see how the space develops
- Documentation
  - can be done in parallel
  - comprehensive tutorials are not feasible with major areas of the engine still deficient
  - docs have a maintenance cost
- Advanced PBR and rendering performance optimizations
  - important and makes for pretty graphics
  - not nearly as blocking for creation of games
  - can continue largely independently from contributors interested in the area
- Input mapping
  - needs more design work
  - not pressing enough yet
- XR (Virtual, augmented and mixed reality) support
  - hard for many to contribute
  - only relevant to a small subset of users
  - can continue in parallel
- Mobile and web support
  - requires tight UI integration to test usability on mobile
  - lifecycle management and stability fixes are important, but need design from experts
  - can continue in parallel
- Animation
  - #1429 (and perhaps #428) needs more user experience reports, docs and testing
  - these should be prototyped in a third-party plugin with MIT + Apache licenses and then brought in-stream once they're more mature
- Networking
  - Bevy is too immature
  - https://github.com/bevyengine/rfcs/pull/19 is the first step towards this

**Reply by Sheepyhead** [reply]
Personally I think UI should be the absolute main focus for 0.7, as it's by far the biggest issue with Bevy, in the sense that it's something literally every developer will need to use sooner or later, and the current state of the API and documentation is not very useful. There's also not really any acceptable alternatives, egui is nice but ultimately mostly useful for debug gui. In my opinion the rest of the notable areas are relevant for 0.7 mostly in terms of what they provide to making the UI system work.

**Reply by TheRawMeatball** [reply]
Hmm, I don't see how input mapping and extended platform support are blocked by UI. FWIW, a solid input mapping system could be a viable thing to focus on, though I don't think it's worth it at this stage.

With platform support, mobile will probably continue to be buggy until lifecycle management is sorted out, but I think a somewhat buggy version could be done without much work - most of the work is already done. And for web, I'm fairly confident it'll solve itself once wgpu releases webgl2 support.

**Reply by inodentry** [reply]
Aren't "system schedule and states" also something that can be done largely in parallel to whatever the official focus is? Also those areas are the domain of @TheRawMeatball and @Ratysz , who can work on it independently from Cart, whenever they are willing and able. :slightly_smiling_face:  Also Cart said he won't spend official focus on ECS. So I don't think "system schedule and states" actually belongs in that first list, but rather the second. :slightly_smiling_face: 

I agree that UI seems to be the most serious blocker for many people's projects as it stands.

However, UI is a very big design area that will probably take a long time to rework (similar to rendering now). IMO It's worth considering whether the work on Assets or Scenes is comparatively straightforward, and if Cart feels like he could make big improvements quickly in those areas, then maybe it's worth spending a few weeks doing those first, before moving the focus to UI. Maybe even cut another release in between (could help get back to the initial plan of having monthly releases :wink: ). It would be a benefit for many users, and get this work out of the way (rather than postponing/deprioritizing it), letting Cart then focus on creating an amazing UI framework in peace :slightly_smiling_face: .

**Reply by alice-i-cecile** [reply]
> I don't see how input mapping and extended platform support are blocked by UI

Right, I should have been clearer there. They're not properly blocked, just heavily interdependent. Input mapping needs a way to unify hot-keys and button presses with the UI. Extended platform support needs touch and widget integration for mobile (and comparable elements for XR); this is hard to properly test / integrate without an existing UI solution.

I've amended the parent post to clarify things <3

**Reply by alice-i-cecile** [reply]
> Aren't "system schedule and states" also something that can be done largely in parallel to whatever the official focus is? Also those areas are the domain of @TheRawMeatball and @Ratysz , who can work on it independently from Cart, whenever they are willing and able. 🙂 Also Cart said he won't spend official focus on ECS. So I don't think "system schedule and states" actually belongs in that first list, but rather the second. 🙂

Hmm, that's very fair. Cart got into it a bit in #2777; I was concerned that it might take too much attention to review the requisite pieces in order to combine it with another release.

**Reply by afonsolage** [reply]
Maybe we can have a list of focus to Cart, Main Devs and community?

I know we can't force community, since folks do what they want, but if we have some directions for folks which wanna help in some spare time, this may help a lot to direct them.

Note: I don't know if bevy already has the concept of core/main Devs, but here I'm referring some nice and notable folks which are contributing to Bevy regualiry and have notable experience and understanding (not my case of course) 😄

**Reply by alice-i-cecile** [reply]
> Maybe we can have a list of focus to Cart, Main Devs and community?

For extensive community-driven efforts, I would probably suggest aiming for a joint RFC or external prototype (like `bevy_kira_audio` is). Animation could badly use this for example, and XR is already collaborating in a very complex PR.

**Reply by alice-i-cecile** [reply]
(Added ray-casting and animation to the topic lists)

**Reply by edgarssilva** [reply]
In my opinion, I think 2d animations, texture atlas and polishing the 2d features so new and inexperienced developers can make something polished. I do think this can bring a lot of new people and maybe can start seeing commercials games popping out (maybe some basic mobile games) but UI should be a big obstacle to that.

**Reply by colepoirier** [reply]
I’d like to build on what @inodentry said about UI potentially taking as long as rendering. In fact I think it may take between 2-4x as long as rendering so I would suggest we come up with a staged approach to the UI rework, where we start building the foundations/data flow first, planning to work up to the higher level things from there, incrementally each release. I’m not sure if what I’m trying to get at is being conveyed clearly so please reply to this so I can better explain it through collaborative discussion with you if you’re interested :)

**Reply by Nilirad** [reply]
#### Concerns about doing UI right now

I prefer **0.7 being a quicker update** than 0.5 and 0.6.
For how I'm very hyped about UI, I still think that bringing it directly on the table will probably extend the development to 0.5/0.6 like times, especially given that **there is not a consensus** on both the implemenation and the API.

If we tackle the basics first, we can then focus efforts into making good decisions about how the UI will be designed and implemented for 0.8.

#### Basics for UI

These are a couple of things I thought about that could be added to the engine as a foundation for UI:

- **A canvas API for drawing shapes**
  - Implementation of bevyengine/rfcs#12 will bring a solid foundation on defining common shapes.
  - We can also learn from [`bevy_prototype_lyon`](https://github.com/Nilirad/bevy_prototype_lyon) about various API strategies, being used by many people for over a year.
- **Assets and boilerplate reduction.** An asset module rework aimed at boilerplate reduction paired with other strategies will probably be very beneficial for UI. For example, now you have to specify everything manually, from fonts to resources about a button state. Proper encapsulation and sane defaults will help a lot to have a cleaner API in less time.

#### My view on 0.7 and beyond timeline

1. _Bevy 0.6 (renderer v2 update) lands_
2. **Assets**, **Canvas API**, **Scenes and reflection** (mostly in parallel; to prepare for UI). Maybe also **Hierarchy cleanup**, since trees are important in UI, and less buggy trees will speed up UI development
3. _Bevy 0.7 (ui v2 prep update) lands_
4. **Editor-Ready UI**
5. _Bevy 0.8 (ui v2 update) lands_
6. From here on, we could focus on the editor, maybe separating preparation and implementation in different releases.

**Reply by inodentry** [reply]
OK, nice to see that there is some community agreement about having a quicker release / shorter dev cycle, to tackle some foundational things, before focusing on UI. I really think this is the best way forward.

So, these are the things we might want to get out of the way, before focusing on UI full-time: 1) most importantly: Assets and Scenes, 2) possibly also Canvas / 2D drawing API (thanks @Nilirad ), and 3) also the shapes/primitives (and i'd like to add the related raycasting stuff) that are sitting in PRs in a close-to-ready state (basically @aevyrie 's stuff). All of these things feel pretty essential for implementing the gameplay and content of many different classes of games.

This way, we can unblock a lot of people's progress on many different kinds of games and ensure we have good foundations. UI will likely take a long time to design and implement properly, so would be good to have all the previously-mentioned stuff out of the way and be able to take our time with it.

**Reply by alice-i-cecile** [reply]
I'm pretty firmly in agreement with @Nilirad's perspective: there's too much foundational work to do for UI still, and we don't have a clear enough direction and vision. A shorter release cycle would help a lot: I wouldn't commit to 0.8 *finally* being the UI release; it may make more sense to split the foundational work even further.

Basic animation is the other foundation that I think we should really try to have in place before UI: anything sophisticated is going to use animation pretty heavily and we need to make sure we can build our UI paradigm around it. As we've discussed with @lassade in #animation-dev, the best way to move forward on this is to get a small group of interested folks to build out a dedicated 3rd-party animation crate that the community can start using and refining ASAP.

**Reply by afonsolage** [reply]
I like the idea of 0.7 being a quicker release, but tbh, I don't know if we should consider this to be a thing, since as stated by Cart already, Bevy is in a heavy design/testing phase and by not having a release schedule, this enable us to do the heavy work needed to Bevy have solid foundations.

Working with smaller realeses may lead to extra work as migration guide, docs updating and others tasks that may take some time.

So I think it's better keep Bevy as "release when it's ready" phase and do what needs to be done, like it was with the new rendering pipeline, so I think UI, and I mean the core/low-level design/API, should be the next big thing, since it's the building blocks for all games and our editor (I know, things like assets, scenes and animation are building blocks too). 

That being said, if Cart feels it's time to have quicker releases and attract more new users to Bevy, so it be.

**Reply by djeedai** [reply]
> Working with smaller releases may lead to extra work as migration guide, docs updating and others tasks that may take some time.

I've opened #2940 before seeing this discussion, and from my limited experience on this project (so bear with me if I misunderstood something) I disagree with the statement, in that there's already apparently consensus (see _e.g._ #2918) that docs etc. are lacking, so I'm not sure the amount of extra work to update those is as large as suggested. There's a non-zero cost for sure, but it sounds to me that the benefits outweigh those costs.

> it's time to have quicker releases and attract more new users to Bevy, so it be.

As mentioned in #2940 on the other hand I'm under the impression that larger chunky releases are indeed more likely to drive away new developers. Having 6 months of bug fixes and improvements not available to new users is not only annoying for those new users (functionalities not working, poorer first impression) but also I think for devs themselves (have to answer again and again issues already fixed, don't see new features they landed used by other user's projects staying on stable releases).

I've led some open-source project before, not of this scale, but with enough visibility to have a few companies use it to ship products. If I can share a bit of personal experience here, as that situation sounds familiar, I feel the biggest mistake we/I made was not releasing often enough, leaving large gaps between releases (6+ months), and eventually driving some users away who where not seeing issues being fixed (because fixed in git but not in release) and were losing interest in the project (for hobbyists) or finding an alternative project (for more serious uses). This is something at the time I didn't necessarily realize, seeing myself the day-to-day improvements which were not visible to users not closely following the project development (that is, consuming from git). There's a motivational component here that I think is important.

> UI: needs heavy-handed decision-making to cut through paralysis

I also wanted to touch on this, from my limited experience so far, so I'm glad to see @alice-i-cecile mentioning it already. I think this is also related to the above, in that there's a motivational component here to lacking Editor and waiting on a larger UI refactor. I've come to observe from other projects, both personally and professionally, that having _something_, even imperfect, is often better than having nothing due to the paralysis of wanting to do "the right thing" (c). I'm the first to fall for the "right thing" approach, so I totally get the arguments for it, and I'm not blaming anyone. I'm not suggesting to rush any decision either and do something at any cost. But the power of having a plugin system means there's probably some room for trial and error, and therefore maybe for short-term actions on those to fill the gaps, even if temporarily until a more mature solution arises?

**Reply by colepoirier** [reply]
@djeedai related to your last point, there is a not very good existing bevy UI, and to your point about the plugin system the current most used solution for better UI is bevy_egui. Cart has expressed that his preference is for third party plugin crates to be developed to explore the space of UI, and eventually aspects of those will be integrated into the official UI. Does this alleviate your concerns with respect to UI? I do think a monthly bug fixes crates.io release would be a potentially better way to go, keeping the large feature addition work in a branch separate from main until it is ready to be merged. This has been discussed in the past and cart expressed his preference for each release to have new features so that it provides more excitement from release announcements (I may be unintentionally misrepresenting his perspective, this is what I remember from discussions months ago), but I would prefer your suggested approach of frequent regular releases. @cart any thoughts?

**Reply by djeedai** [reply]
@colepoirier yes and no. It's good that `bevy_egui` provides an alternative to the built-in UI module. And crates like `bevy-inspector-egui` make good use of it to provide what after 5 min of use looks to me like an invaluable tool to help development. However I don't think an immediate-mode GUI is suited for an Editor. Unity3D invested heavily in immediate-mode UI for their editor, and then had to back off and invest even more heavily trying to push people away from it, due to performance concerns.

If the UI is blocking the Editor, it sounds to me like exploring with an Editor-focused UI plugin would unblock that, and therefore also allow exploring building an Editor itself on top of it right away in parallel. And waiting on "the perfect solution" for UI to start an Editor might take too much time for some folks to stick around, and/or drive away new users. As @alice-i-cecile said (and that's how I understand the statement, sorry if I'm misreading it) some "heavy-handed decision-making to cut through paralysis".

**Reply by colepoirier** [reply]
@djeedai I think you are misunderstanding a bit. The immediate mode debugging tooling is just that, its not official bevy UI, there is no investment in it by the bevy project whatsoever. I’m not sure how “exploring with an Editor-focused UI plugin would unblock” the editor or the UI, could you expand on what you mean? The heavy handed decision making has to be made by cart, we’re waiting on cart for the big UI rewrite and the editor will come after that, it can’t happen in parallel because it requires the UI solution to exist first.

**Reply by Barugon** [reply]
> * UI
>     * needs heavy-handed decision-making to cut through paralysis
>     * some serious open design questions, particularly around the data flow model
>     * heavy user impact: this is one of the most serious challenges with making Bevy games right now

This.

**Reply by colepoirier** [reply]
What about ‘this’ @Barugon?

**Reply by Barugon** [reply]
> What about ‘this’ @Barugon?

My vote for what should be a primary focus for 0.7.

**Reply by colepoirier** [reply]
Ah gotcha thanks for this!

**Comment by alice-i-cecile** [comment]
# Engine development process improvements

Again, topics for discussion. A lot of this follows up on the work started by @NathanSWard in #2256.

1. Project roadmap: how do we want to communicate this?
2. Issue and PR lifecycle management using Github boards.
3. More granular PR status labels: `needs-review` in particular would be very useful.
4. PR review process changes to reduce merge latency?
5. Minor releases for Bevy?
6. Focus areas meta-discussion: how do we set, stick to and communicate priorities?

**Reply by Nilirad** [reply]
>Minor releases for Bevy?

Assuming that we will adopt a shorter release cycle for the next versions, I don't think it is a good idea. Even small improvements to the user experience should be a new major release, just like the first versions of Bevy. It will be a different story once we hit 1.0, but it's too early to talk about it now. 

>Focus areas meta-discussion: how do we set, stick to and communicate priorities?

I think Cart should have the exclusive right to ultimately define the focus areas of Bevy. Obviously anyone can express their thoughts but this has to happen naturally, without meetings or discussions about what the next focus area should be. The communication can happen through the #dev-announcements channel and/or as a GitHub tracker issue (like the current three focus areas are already defined). I think we're already sticking to the focus areas, by giving merge priority to pertinent PRs.

**Reply by Weibye** [reply]
3. A `S-Needs-Review` would indeed be helpful. Do we also need a `S-Need-Author-Feedback / Response` ?

**Reply by alice-i-cecile** [reply]
There's a bit of justified concern about how actionable that would be. The label is mostly useless to everyone but the single author.

**Comment by colepoirier** [comment]
Building on your comments @alice-i-cecile, I’d like to suggest that there are three primary focus areas that need to be worked on before UI, and three smaller, more independent areas that can proceed entirely in parallel. I’ve identified what I think can be led primarily by community members, ones where I think @cart need to be equally involved, and the ones which I predict cart should be the primary driver.

Primary
1. Ray-casting, picking and geometric primitives (more community, less cart)
2. Scenes, reflection, assets (cart)
3. Animation (cart)

Smaller, secondary, parallel tasks
1. Mobile and web support (mostly community)
2. Hierarchy clean-up (Cart)
3. System schedule and states (@Ratysz and/or @TheRawMeatball and some Cart)

I would also like to suggest that Cart start working on UI, primarily participating in discussion starting after 0.6 is released, in parallel with whatever his main focus is for the next 3-4 releases to start building out plans and implementing prototypes of foundational UI components. I think that this needs to happen in order for us to have any hope of actually “getting to UI” because it is simply too huge a domain to do all at once. I think if cart starts working lightly on UI at first by just participating in discussions, and builds up his involvement/focus/efforts on ui over the releases of the other focus areas, that this will be the best way to slowly work up to doing the full UI focus eventually.

What do you think?

**Reply by alice-i-cecile** [reply]
The hierarchy cleanup can comfortably be tackled without much Cart involvement. Relations would be one approach to solve it, but the scope is too large and uncertain for me to recommend it as a prerequisite.

But doing a code quality / API pass would fix enough to make it usable for UI and is fundamentally mundane engineering work.

**Reply by alice-i-cecile** [reply]
I don't think the first item is necessary for UI.

A canvas API would be more useful for it, but not essential immediately: rectangles go a long way. Geometric primitives would probably be useful for that, but as seen in 'bevy_prototype_lyon`, not truly needed.

**Reply by colepoirier** [reply]
I should clarify that I don’t think all of these are prerequisites for UI, but that it seems like in terms of timeframes they could be completed before the big UI focus especially since a lot of them can be done without cart. I think that the UI discussion and early development process could take a long time so it should start with low but regular time commitments from cart.

I agree that animation and canvas would be a great thing to get done together because of how useful I’m assuming that will be to 2D. And I would say getting 2D capabilities like that implemented would be a big win for the engines maturity overall.

**Reply by inodentry** [reply]
I quite like this list, though not necessarily as "a list of prerequisites for UI", but rather "a list of other foundational work with big user impact that would be nice to complete first regardless of UI". It would do wonders for the maturity of the engine and enabling people to work on more different kinds of games and more ambitious projects. (yes, arguably UI itself is more of a blocker for many projects, but it would also take a much longer time to get ready)

**Reply by afonsolage** [reply]
> I quite like this list, though not necessarily as "a list of prerequisites for UI", but rather "a list of other foundational work with big user impact that would be nice to complete first regardless of UI". It would do wonders for the maturity of the engine and enabling people to work on more different kinds of games and more ambitious projects. (yes, arguably UI itself is more of a blocker for many projects, but it would also take a much longer time to get ready)

Sorry if I misunderstood, but what's the point of having  people to work on more different kinds of games and more ambitious projects, if there is no UI?

**Comment by Joe23232** [comment]
I was thinking we had a similar UI like [DearPyGUI](https://github.com/hoffstadt/DearPyGui#readme)

![image](https://user-images.githubusercontent.com/34926497/144388298-5fcce781-8520-40d5-8a58-91ff611cc24f.png)

And we use similar syntax as to theirs as it is quite simple by the looks of things.

What do you guys say?

**Reply by colepoirier** [reply]
I have good news for you, we have this already, it is called bevy_egui!

**Reply by Joe23232** [reply]
@colepoirier But that is immediate, not retained unlike Dear PyGui.

**Reply by colepoirier** [reply]
The api is the same regardless.
